smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/metrics/MetricDecorator.java

```diff
@@ -0,0 +1,39 @@
+package io.smallrye.reactive.messaging.metrics;
+
+import java.util.function.Consumer;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Instance;
+import javax.inject.Inject;
+
+import org.eclipse.microprofile.metrics.Counter;
+import org.eclipse.microprofile.metrics.MetricRegistry;
+import org.eclipse.microprofile.metrics.Tag;
+import org.eclipse.microprofile.reactive.messaging.Message;
+import org.eclipse.microprofile.reactive.streams.operators.PublisherBuilder;
+
+import io.smallrye.reactive.messaging.PublisherDecorator;
+
+@ApplicationScoped
+public class MetricDecorator implements PublisherDecorator {
+
+    @Inject
+    private Instance<MetricRegistry> registryInstance;
```

Comment:
I presume we're using `Instance` here so we can lazy load?

Any particular advantage over just grabbing it from the start?

Reply:
I was using an `Instance` here so that it would be able to run whether or not the `MetricRegistry` is present, on the basis that someone might want to use reactive messaging without metrics at all.

Reply:
Would you get a CNFE when `MetricsDecorator` is instantiated if metrics were not present?

Though an instance of it may not be there, it still needs the Type to be able to create the `Instance`, or am I wrong?

Reply:
You're right, `MetricDecorator` needs the `MetricRegistry` class to be loadable, so it will always require that the Metrics API is present, and I've added a compile dependency on the Metrics API to the pom for the provider.

However, as it's written, it doesn't require an MP metrics implementation to be present to provide the `MetricsRegistry` implementation and injectable bean.

Note: If we didn't want the provider to have a dependency on the metrics API, we could move the `MetricDecorator` to a separate module.

Reply:
Sorry, missed the dependency when looking at this.

Thanks